src/bus/peer.c

```diff
@@ -449,6 +502,26 @@ int peer_add_match(Peer *peer, const char *rule_string) {
         return 0;
 }
 
+static Name *peer_match_rule_to_name(MatchRule *rule) {
+        if (!rule->keys.sender)
+                return NULL;
+        /*
+         * A match can be associated with a name in two cases:
+         *  - if it is a NameOwnerChanged subscription on the name, or
+         *  - if it is a sender match on the name.
+         */
+        if (strcmp(rule->keys.sender, "org.freedesktop.DBus") == 0) {
+                if (rule->keys.filter.member && strcmp(rule->keys.filter.member, "NameOwnerChanged") == 0 &&
+                    rule->keys.filter.args[0] && strcmp(rule->keys.filter.args[0], "org.freedesktop.DBus") != 0 &&
+                    rule->keys.filter.args[0][0] != ':')
+                        return c_container_of(rule->registry, Name, name_owner_changed_matches);
+        } else if (rule->keys.sender[0] != ':') {
+                return c_container_of(rule->registry, Name, sender_matches);
+        }
```

Comment:
Yeah... This needs a big fat `XXX: WTF?`

I am fine with it for now, since we had the same conceptual issue before. But this is not cool. Maybe add a field to `rule` that pins the name. Or a boolean at least.